### ISSUE TITLE
TokenCollector crashes 

### ISSUE THREAD

**Issue by hokein** [clangd/clangd#328]
open the file `llvm-project/clang/test/Sema/pragma-pipeline.cpp` in your editor, clangd will crash.

crash stack:

```
clangd: llvm-project/clang/lib/Tooling/Syntax/Tokens.cpp:553: void clang::syntax::TokenCollector::Builder::processExpandedToken(unsigned int &): Assertion `File.SpelledTokens[NextSpelled[FID]].location() == L && "no corresponding token in the spelled stream"' failed.
I[09:02:54.005] Loaded compilation database from llvm-project/build
 #0 0x00000000004e8dcf llvm::sys::PrintStackTrace(llvm::raw_ostream&) llvm-project/llvm/lib/Support/Unix/Signals.inc:564:13
 #1 0x00000000004e6ff2 llvm::sys::RunSignalHandlers() llvm-project/llvm/lib/Support/Signals.cpp:69:18
 #2 0x00000000004e911c SignalHandler(int) llvm-project/llvm/lib/Support/Unix/Signals.inc:396:3
 #3 0x00007f297b85e520 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x13520)
 #4 0x00007f297b32e081 raise /build/glibc-QG9qOO/glibc-2.29/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f297b319535 abort /build/glibc-QG9qOO/glibc-2.29/stdlib/abort.c:81:7
 #6 0x00007f297b31940f _nl_load_domain /build/glibc-QG9qOO/glibc-2.29/intl/loadmsgcat.c:1177:9
 #7 0x00007f297b326b92 (/lib/x86_64-linux-gnu/libc.so.6+0x32b92)
 #8 0x000000000096ddc8 (llvm-project/build/bin/clangd+0x96ddc8)
 #9 0x000000000096d51b clang::syntax::TokenCollector::Builder::build() && llvm-project/clang/lib/Tooling/Syntax/Tokens.cpp:515:64
#10 0x000000000096d3fd llvm::DenseMap<int, clang::SourceLocation, llvm::DenseMapInfo<int>, llvm::detail::DenseMapPair<int, clang::SourceLocation> >::~DenseMap() llvm-project/llvm/include/llvm/ADT/DenseMap.h:725:23
#11 0x000000000096d3fd clang::syntax::TokenCollector::Builder::~Builder() llvm-project/clang/lib/Tooling/Syntax/Tokens.cpp:493:0
#12 0x000000000096d3fd clang::syntax::TokenCollector::consume() && llvm-project/clang/lib/Tooling/Syntax/Tokens.cpp:702:0
#13 0x0000000000a6cf99 std::_Vector_base<clang::Decl*, std::allocator<clang::Decl*> >::_Vector_impl_data::_Vector_impl_data(std::_Vector_base<clang::Decl*, std::allocator<clang::Decl*> >::_Vector_impl_data&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:100:17
#14 0x0000000000a6cf99 std::_Vector_base<clang::Decl*, std::allocator<clang::Decl*> >::_Vector_impl::_Vector_impl(std::_Vector_base<clang::Decl*, std::allocator<clang::Decl*> >::_Vector_impl&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:141:0
#15 0x0000000000a6cf99 std::_Vector_base<clang::Decl*, std::allocator<clang::Decl*> >::_Vector_base(std::_Vector_base<clang::Decl*, std::allocator<clang::Decl*> >&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:305:0
#16 0x0000000000a6cf99 std::vector<clang::Decl*, std::allocator<clang::Decl*> >::vector(std::vector<clang::Decl*, std::allocator<clang::Decl*> >&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:569:0
#17 0x0000000000a6cf99 clang::clangd::(anonymous namespace)::ClangdFrontendAction::takeTopLevelDecls() llvm-project/clang-tools-extra/clangd/ParsedAST.cpp:94:0
#18 0x0000000000a6cf99 clang::clangd::ParsedAST::build(llvm::StringRef, std::unique_ptr<clang::CompilerInvocation, std::default_delete<clang::CompilerInvocation> >, llvm::ArrayRef<clang::clangd::Diag>, std::shared_ptr<clang::clangd::PreambleData const>, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, clang::clangd::SymbolIndex const*, clang::clangd::ParseOptions const&) llvm-project/clang-tools-extra/clangd/ParsedAST.cpp:405:0
#19 0x0000000000a6f1e3 clang::clangd::buildAST(llvm::StringRef, std::unique_ptr<clang::CompilerInvocation, std::default_delete<clang::CompilerInvocation> >, llvm::ArrayRef<clang::clangd::Diag>, clang::clangd::ParseInputs const&, std::shared_ptr<clang::clangd::PreambleData const>) llvm-project/clang-tools-extra/clangd/ParsedAST.cpp:553:10
#20 0x0000000000ac5fb1 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:729:6
#21 0x0000000000ac5fb1 std::__shared_ptr<clang::clangd::PreambleData const, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:1169:0
#22 0x0000000000ac5fb1 clang::clangd::(anonymous namespace)::ASTWorker::update(clang::clangd::ParseInputs, clang::clangd::WantDiagnostics)::$_0::operator()() llvm-project/clang-tools-extra/clangd/TUScheduler.cpp:505:0
#23 0x0000000000ac5fb1 void llvm::unique_function<void ()>::CallImpl<clang::clangd::(anonymous namespace)::ASTWorker::update(clang::clangd::ParseInputs, clang::clangd::WantDiagnostics)::$_0>(void*) llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:157:0
#24 0x0000000000ac22f7 clang::clangd::(anonymous namespace)::ASTWorker::run() llvm-project/clang-tools-extra/clangd/TUScheduler.cpp:763:5
#25 0x0000000000ac22f7 clang::clangd::(anonymous namespace)::ASTWorker::create(llvm::StringRef, clang::clangd::GlobalCompilationDatabase const&, clang::clangd::TUScheduler::ASTCache&, clang::clangd::AsyncTaskRunner*, clang::clangd::Semaphore&, clang::clangd::DebouncePolicy, bool, clang::clangd::ParsingCallbacks&)::$_5::operator()() const llvm-project/clang-tools-extra/clangd/TUScheduler.cpp:345:0
#26 0x0000000000ac22f7 void llvm::unique_function<void ()>::CallImpl<clang::clangd::(anonymous namespace)::ASTWorker::create(llvm::StringRef, clang::clangd::GlobalCompilationDatabase const&, clang::clangd::TUScheduler::ASTCache&, clang::clangd::AsyncTaskRunner*, clang::clangd::Semaphore&, clang::clangd::DebouncePolicy, bool, clang::clangd::ParsingCallbacks&)::$_5>(void*) llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:157:0
#27 0x0000000000ab8c2a llvm::PointerIntPair<llvm::PointerUnion<llvm::unique_function<void ()>::TrivialCallback*, llvm::unique_function<void ()>::NonTrivialCallbacks*>, 1u, bool, llvm::PointerLikeTypeTraits<llvm::PointerUnion<llvm::unique_function<void ()>::TrivialCallback*, llvm::unique_function<void ()>::NonTrivialCallbacks*> >, llvm::PointerIntPairInfo<llvm::PointerUnion<llvm::unique_function<void ()>::TrivialCallback*, llvm::unique_function<void ()>::NonTrivialCallbacks*>, 1u, llvm::PointerLikeTypeTraits<llvm::PointerUnion<llvm::unique_function<void ()>::TrivialCallback*, llvm::unique_function<void ()>::NonTrivialCallbacks*> > > >::PointerIntPair() llvm-project/llvm/include/llvm/ADT/PointerIntPair.h:48:12
#28 0x0000000000ab8c2a llvm::unique_function<void ()>::unique_function(std::nullptr_t) llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:174:0
#29 0x0000000000ab8c2a clang::clangd::AsyncTaskRunner::runAsync(llvm::Twine const&, llvm::unique_function<void ()>)::$_1::operator()() llvm-project/clang-tools-extra/clangd/Threading.cpp:95:0
#30 0x0000000000ab8c2a void llvm::unique_function<void ()>::CallImpl<clang::clangd::AsyncTaskRunner::runAsync(llvm::Twine const&, llvm::unique_function<void ()>)::$_1>(void*) llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:157:0
```

**Comment by sam-mccall** [comment]
Coincidentally I was rewriting some of this to simplify and prepare a fix for another bug (D77615).

I'll minimize and see if the new impl fixes it

**Comment by sam-mccall** [comment]
Minimized:
```
void test() {
  #pragma clang loop_pipeline_initiation_interval(0)
}
```

Expanded token sequence is bad:
```
Token(`void`, void, length = 4)
Token(`test`, identifier, length = 4)
Token(`(`, l_paren, length = 1)
Token(`)`, r_paren, length = 1)
Token(`{`, l_brace, length = 1)
Token(`0`, numeric_constant, length = 1)
Token(``, eof, length = 0)
```

**Comment by gislan** [comment]
Looks like this was fixed by Sam in d66afd6dde542dc373f87e07fe764c071fe20d76
